proposals/unions.md

```diff
@@ -0,0 +1,322 @@
+# Unions
+
+Champion issue: https://github.com/dotnet/csharplang/issues/8928
+
+## Summary
+[summary]: #summary
+
+*Unions* is a set of interlinked features, that combine to provide C# support for union types:
+
+- *Union types*: Classes and structs that implement the `IUnion` interface are recognized as *union types*, and support the *union behaviors*.
+- *Case types*: Union types have a set of *case types*, which is given by the set of public instance constructors with exactly one parameter. The type of each such constructor parameter is a case type for the union type.
+- *Union behaviors*: All union types support the following *union behaviors*:
+    - *Union conversions*: There are implicit *union conversions* from each case type to a union type.
+    - *Union matching*: Pattern matching against union values implicitly "unwraps" their contents, applying the pattern to the underlying value instead.
+    - *Union exhaustiveness*: Switch expressions over union values are exhaustive when all case types have been matched, without need for a fallback case.
+    - *Union nullability*: Nullability analysis has enhanced tracking of the null state of a union's contents.
+- *Union patterns*: All union types follow a basic *union pattern* (implement `IUnion` and offer public constructors for case types), but there are additional optional patterns that enable the union behaviors to be implemented in ways that may be more efficient.
+- *Union declarations*: A shorthand syntax allows declaration of union types directly. The implementation is "opinionated" - a struct declaration that follows the basic union pattern and stores the contents as a single reference field.
+- *Union interfaces*: A few interfaces are known by the language and used in its implementation of unions. `IUnion` is used to recognize union types and support union matching. An optional interface, `IUnion<TUnion>`, adds broader support for creating union values from potential case types, and is implemented by the union types arising from union declarations.
+
+## Motivation
+[motivation]: #motivation
+
+Unions are a long-requested C# feature, which allows expressing values from a closed set of types in a way that pattern matching can trust to be exhaustive.
+
+The separation between union *types* and union *declarations* allows C# to have a succinct union declaration syntax with opinionated semantics, while also allowing existing types or types with other implementation choices to opt into union behaviors.
+
+The proposed unions in C# are unions of *types* and not "discriminated" or "tagged". "Discriminated unions" can be expressed in terms of "type unions" by using fresh type declarations as case types.
```

Comment:
maybe link to an explanatory section taht shows how one would do this.  It would make sense to go later on to demontrate how to build a discriminated union from the union pieces we are spec'ing out.

Reply:
I added examples later. I don't want to link here, since the section doesn't link forward to anything else it mentions.